### PR TITLE
LLEXT: automatic gcc / xt-clang switching

### PR DESCRIPTION
**Pull request by lyakh**
We build different LLEXT object types with gcc and with xt-cland and switching between them should be automatic

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]

File: scripts/xtensa-build-zephyr.py [hunk 1]
```diff
@@ -809,6 +809,13 @@ def build_platforms():
 		if args.debug:
 			overlays.append(str(pathlib.Path(SOF_TOP, "app", "debug_overlay.conf")))
 
+		# The xt-cland Cadence toolchain currently cannot link shared
+		# libraries for Xtensa. Therefore when it's used we switch to
+		# building relocatable ELF objects.
+		if ("ZEPHYR_TOOLCHAIN_VARIANT" in platf_build_environ.keys() and
+		    platf_build_environ["ZEPHYR_TOOLCHAIN_VARIANT"] == 'xt-clang'):
```
Comment (marc-hb) [inline scripts/xtensa-build-zephyr.py:816]: Python can already do that for you:
```suggestion
if platf_build_environ.get("ZEPHYR_TOOLCHAIN_VARIANT") == 'xt-clang'
```

File: scripts/llext_link_helper.py [hunk 2]
```diff
@@ -32,56 +33,107 @@ def parse_args():
 def main():
 	parse_args()
 
+	TYPE_NONE = 0
+	TYPE_DATA = 1
+	TYPE_BSS = 2
+	last_type = TYPE_NONE
+
+	SHF_ALLOC = 2
+	SHF_EXECINSTR = 4
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:42]: A simple `git grep` showed that these are already defined, no need to duplicate

```
ipython3
In [1]: from elftools.elf.constants import SH_FLAGS

In [2]: SH_FLAGS.SHF_EXECINSTR
Out[2]: 4
```

File: scripts/llext_link_helper.py [hunk 3]
```diff
@@ -32,56 +33,107 @@ def parse_args():
 def main():
 	parse_args()
 
+	TYPE_NONE = 0
+	TYPE_DATA = 1
+	TYPE_BSS = 2
+	last_type = TYPE_NONE
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:39]: https://docs.python.org/3/library/enum.html

```python
from enum import Enum

SECTION_TYPE = Enum('SECTION_TYPE', ['DATA', 'BSS'])

Don't define TYPE_NONE (the "billion dollar mistake"), just use Python's `None`.
```

File: scripts/llext_link_helper.py [hunk 4]
```diff
@@ -32,56 +33,107 @@ def parse_args():
...
 
 	text_addr = int(args.text_addr, 0)
-	p = re.compile(r'(^lib|\.so$)')
-	module = p.sub('', args.file)
 
-	if elf.num_segments() != 0:
-		# A shared object type image, it contains segments
-		sections = ['.text', '.rodata', '.data', '.bss']
-		alignment = [0x1000, 0x1000, 0x10, 0x1000]
+	# File names differ when building shared or relocatable objects
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:48]: Does this really belong to this script? As opposed to its caller.

File: scripts/llext_link_helper.py [hunk 5]
```diff
@@ -32,56 +33,107 @@ def parse_args():
...
+	module = p.sub('', fname)
 
 	last_increment = 0
 
-	command = [args.command]
+	command = [args.command, f'-Wl,-Ttext=0x{text_addr:x}']
+	data_found = False
+	bss_found = False
+
+	# The layout seems to be: .text, followed by other .data, .rodata and
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:64]: "seems to be"?

File: scripts/llext_link_helper.py [hunk 6]
```diff
@@ -32,56 +33,107 @@ def parse_args():
...
+		s_size = section.header['sh_size']
+		s_offset = section.header['sh_offset']
+
+		if (s_flags & (SHF_ALLOC | SHF_EXECINSTR) == SHF_ALLOC | SHF_EXECINSTR and
+		    s_type == 'SHT_PROGBITS'):
+			if s_name != '.text':
+				print(f"Warning! Non-standard executable section {s_name}")
+
+			if bss_found or data_found:
+				print("ERROR: So far .text is expected first. Fix the script!")
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:80]: which script? This one?

File: scripts/llext_link_helper.py [hunk 7]
```diff
@@ -32,56 +33,107 @@ def parse_args():
...
+				return
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:81]: ERROR == success?

File: scripts/llext_link_helper.py [hunk 8]
```diff
@@ -32,56 +33,107 @@ def parse_args():
...
+
+			# We have already inserted a command
+			continue
+
+		if s_flags & SHF_ALLOC and s_type == 'SHT_PROGBITS':
+			# .data, .rodata or similar
+			if last_type == TYPE_BSS and data_found:
+				# Trouble: intermixed data and bss
+				print('ERROR: data again after data and .bss! Aborting')
+				return
```
Comment (marc-hb) [inline scripts/llext_link_helper.py:93]: Aborting with a success code?

Maybe:

```
assert last_type != TYPE_BSS or not data_found, "data again after data and .bss! Aborting'
```

File: scripts/llext_link_helper.py (hunk 2 above)
Comment (lyakh) [reply scripts/llext_link_helper.py:42]: > In [1]: from elftools.elf.constants import SH_FLAGS

that's the line I couldn't find, thanks

File: scripts/llext_link_helper.py (hunk 4 above)
Comment (lyakh) [reply scripts/llext_link_helper.py:48]: we do deal with both cases internally, and passing a module name additionally to the file name would be redundant

File: scripts/llext_link_helper.py (hunk 5 above)
Comment (lyakh) [reply scripts/llext_link_helper.py:64]: @marc-hb I haven't found anywhere a statement, that that should be the case, but so far that's what I've seen

File: scripts/llext_link_helper.py (hunk 6 above)
Comment (lyakh) [reply scripts/llext_link_helper.py:80]: @marc-hb yes

Comment (marc-hb) [reply scripts/llext_link_helper.py:80]: Then the usual "FIXME"

```suggestion
			assert not bss_found and not data_found, "FIXME: So far .text is expected first"
```

File: scripts/llext_link_helper.py (hunk 5 above)
Comment (marc-hb) [reply scripts/llext_link_helper.py:64]: I haven't tried to understand the logic so I wonder why it matters. Can't you just perform two passes? (1) collect all required information, no side-effect (2) do what must be done.

**Review by marc-hb — COMMENTED** [review]
Code style looks good to me.

Someone must review the actual ELF logic.

**Review by kv2019i — APPROVED** [review]

**Comment by kv2019i** [comment]
@lyakh  Ready to go, pending mandatory CI to pass

**Comment by lyakh** [comment]
@wszypelt is quickbuild hanging here, can we restart, please?

**Review by marc-hb — COMMENTED** [review]
When you combine this `llext_relocatable.conf` hack with the other, `repro-build.conf` hack in 87e973d32f4e, we end up in a situation where no CI ever tests the default configuration, which currently prints this really weird "undefined/n" LLEXT warning that I have never seen before:

```
warning: LIBRARY_MANAGER  was assigned the value 'y' but
got the value 'n'. Check these unsatisfied dependencies: CONFIG_LLEXT (=n).

grep LLEXT ../build-mtl/zephyr/.config
../build-mtl/zephyr/.config:CONFIG_LLEXT=y # !!!!

ninja -C ../build-mtl menuconfig

 config LIBRARY_MANAGER
        bool "Library Manager Support"
        default n
        depends on CONFIG_LLEXT(undefined/n)
```

@lyakh can you explain this Kconfig LLEXT mystery?


Good example of:
- #9386

**Comment by marc-hb** [comment]
Even with `llext_relocatable.conf`, another Kconfig warning:

https://sof-ci.01.org/sofpr/PR9403/build7332/build/firmware_community/mtl.log
```
warning: COMP_MULTIBAND_DRC (defined at
/srv/home/jenkins/workspace/sof_config_build@2/sof/zephyr/../src/audio/multiband_drc/Kconfig:3) was
assigned the value 'y' but got the value 'n'. Check these unsatisfied dependencies: (COMP_DRC = y)
```

**Pull request merged**